Update to Flask ≥ 1.0

ipwb currently uses Flask version 0.12.4. #613 was an attempt by a bot to update the repo with a newer version to solve a (still present) security issue. This caused ipwb tests to fail due to some difference between 0.12.3 (at the time) and 1.0.

This ticket represents an effort to move to a newer version of Flask.

Within `queryTimeGate()` of replay.py, the code:

```
if not ipwbUtils.isRFC1123Compliant(adt):
        abort(400)
```

is properly reached and executed. A 400 is returned per the test:

```
400 Bad Request: The browser (or proxy) sent a request that this server could not understand.
```

yet the test fails:
```
warc = '5mementos.warc', lookup = 'timegate/memento.us'
acceptdatetime = 'Thu, 31 May 2007 20:35:00', status = 400

    @pytest.mark.parametrize("warc,lookup,acceptdatetime,status", [
        ('5mementos.warc', 'timegate/memento.us',
         'Thu, 31 May 2007 20:35:00 GMT', 302),
        ('5mementos.warc', 'timegate/memento.us',
         'Thu, 31 May 2007 20:35:00', 400),
        ('5mementos.warc', 'timegate/memento.us',
         'Thu, 31 May 2007 20:35 GMT', 400),
        ('5mementos.warc', 'timegate/memento.us',
         '20181001123636', 400)
    ])
    def test_acceptdatetime_status(warc, lookup, acceptdatetime, status):
        ipwbTest.startReplay(warc)

        headers = {'Accept-Datetime': acceptdatetime}

        resp = requests.get('http://localhost:5000/{}'.format(lookup),
                            allow_redirects=False, headers=headers)
>       assert resp.status_code == status
E       assert 500 == 400
E        +  where 500 = <Response [500]>.status_code
```
Why is this 500 and not 400, as returned from the code?

--- a/Dockerfile
+++ b/Dockerfile
@@ -61,7 +61,7 @@ RUN        pip install -r test-requirements.txt
 COPY       . ./
 ARG        SKIPTEST=false
 RUN        $SKIPTEST || pycodestyle
-RUN        $SKIPTEST || (ipfs daemon & while ! curl -s localhost:5001 > /dev/null; do sleep 1; done && py.test --cov=./)
+RUN        $SKIPTEST || (ipfs daemon & while ! curl -s localhost:5001 > /dev/null; do sleep 1; done && py.test -s --cov=./)
 
 
 # Final production image

--- a/ipwb/replay.py
+++ b/ipwb/replay.py
@@ -271,8 +271,7 @@ def showMemento(urir, datetime):
     else:
         resp = show_uri(uri, newDatetime)
 
-    resp.headers.add('Link', linkHeader)
-    # resp.headers['Link'] = linkHeader
+    resp.headers['Link'] = linkHeader
 
     return resp
 

--- a/ipwb/replay.py
+++ b/ipwb/replay.py
@@ -271,7 +271,8 @@ def showMemento(urir, datetime):
     else:
         resp = show_uri(uri, newDatetime)
 
-    resp.headers['Link'] = linkHeader
+    resp.headers.add('Link', linkHeader)
+    # resp.headers['Link'] = linkHeader
 
     return resp
 
@@ -334,7 +335,7 @@ def queryTimeGate(urir):
         adt = ipwbUtils.getRFC1123OfNow()
 
     if not ipwbUtils.isRFC1123Compliant(adt):
-        abort(400)
+        return "Bad Request", 400
 
     datetime14 = ipwbUtils.rfc1123ToDigits14(adt)
 
@@ -652,15 +653,15 @@ def show_uri(path, datetime=None):
         def handler(signum, frame):
             raise HashNotFoundError()
 
-        if os.name != 'nt':  # Bug #310
-            signal.signal(signal.SIGALRM, handler)
-            signal.alarm(10)
+        # if os.name != 'nt':  # Bug #310
+        #    signal.signal(signal.SIGALRM, handler)
+        #    signal.alarm(10)
 
         payload = IPFS_API.cat(digests[-1])
         header = IPFS_API.cat(digests[-2])
 
-        if os.name != 'nt':  # Bug #310
-            signal.alarm(0)
+        # if os.name != 'nt':  # Bug #310
+        #    signal.alarm(0)
 
     except ipfsapi.exceptions.TimeoutError:
         print("{0} not found at {1}".format(cdxjParts[0], digests[-1]))
@@ -671,23 +672,23 @@ def show_uri(path, datetime=None):
     except TypeError as e:
         print('A type error occurred')
         print(e)
-        abort(500)
+        return "A Type Error Occurred", 500
     except HTTPError as e:
         print("Fetching from the IPFS failed")
         print(e)
-        abort(503)
+        return "Fetching from IPFS failed", 503
     except HashNotFoundError:
         if payload is None:
             print("Hashes not found:\n\t{0}\n\t{1}".format(
                 digests[-1], digests[-2]))
-            abort(404)
+            return "Hashed not found", 404
         else:  # payload found but not header, fabricate header
             print("HTTP header not found, fabricating for resp replay")
             header = ''
     except Exception as e:
         print('Unknown exception occurred while fetching from ipfs.')
         print(e)
-        abort(500)
+        return "An unknown exception occurred", 500
 
     if 'encryption_method' in jObj:
         keyString = None

--- a/requirements.txt
+++ b/requirements.txt
@@ -1,6 +1,6 @@
 warcio>=1.5.3
 ipfshttpclient>=0.4.12
-Flask==0.12.4
+Flask==1.1.1
 pycryptodome>=3.4.11
 requests>=2.19.1
 beautifulsoup4>=4.6.3

--- a/setup.py
+++ b/setup.py
@@ -25,7 +25,7 @@ setup(
     install_requires=[
         'warcio>=1.5.3',
         'ipfshttpclient>=0.4.12',
-        'Flask==0.12.4',
+        'Flask==1.1.1',
         'pycryptodome>=3.4.11',
         'requests>=2.19.1',
         'beautifulsoup4>=4.6.3',

